### PR TITLE
feat(workspace): add a new datasource to querry workspace flavors

### PR DESCRIPTION
**Pull request by ruwenqiang123**
**What this PR does / why we need it**:
Add datasource huaweicloud_workspace_flavors, used for querying the workspace ProductList
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud# make testacc TEST="./huaweicloud/services/acceptance/workspace" TESTARGS="-run TestAccDatasourceFlavors_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run TestAccDatasourceFlavors_basic -timeout 360m -parallel 4
=== RUN   TestAccDatasourceFlavors_basic
=== PAUSE TestAccDatasourceFlavors_basic
=== CONT  TestAccDatasourceFlavors_basic
--- PASS: TestAccDatasourceFlavors_basic (17.42s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 17.470s

```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud# make testacc TEST="./huaweicloud/services/acceptance/workspace" TESTARGS="-run TestAccDataSourceFlavors_filterByMemory"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run TestAccDataSourceFlavors_filterByMemory -timeout 360m -parallel 4
=== RUN   TestAccDataSourceFlavors_filterByMemory
=== PAUSE TestAccDataSourceFlavors_filterByMemory
=== CONT  TestAccDataSourceFlavors_filterByMemory
--- PASS: TestAccDataSourceFlavors_filterByMemory (14.13s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 14.185s
```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud# make testacc TEST="./huaweicloud/services/acceptance/workspace" TESTARGS="-run TestAccDataSourceFlavors_filterByOsType"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run TestAccDataSourceFlavors_filterByOsType -timeout 360m -parallel 4
=== RUN   TestAccDataSourceFlavors_filterByOsType
=== PAUSE TestAccDataSourceFlavors_filterByOsType
=== CONT  TestAccDataSourceFlavors_filterByOsType
--- PASS: TestAccDataSourceFlavors_filterByOsType (12.26s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 12.295s
```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud# make testacc TEST="./huaweicloud/services/acceptance/workspace" TESTARGS="-run TestAccDataSourceFlavors_filterByAvailabilityZone"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run TestAccDataSourceFlavors_filterByAvailabilityZone -timeout 360m -parallel 4
=== RUN   TestAccDataSourceFlavors_filterByAvailabilityZone
=== PAUSE TestAccDataSourceFlavors_filterByAvailabilityZone
=== CONT  TestAccDataSourceFlavors_filterByAvailabilityZone
--- PASS: TestAccDataSourceFlavors_filterByAvailabilityZone (12.24s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 12.294s
```

### CODE REVIEW

File: docs/data-sources/workspace_flavors [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+* `memory` - (Optional, Int) Specifies the memory size of the workspace flavors.
+
+* `os_type` - (Optional, String) The operating system type of the workspace flavors, currently supported: Windows, Linux.  
+  The options are as follows:
+    - **Windows**: The operating system type of the workspace flavors is Windows.
+    - **Linux**: The operating system type of the workspace flavors is Linux.
+
+* `availability_zone` - (Optional, String) Available partition of the workspace flavors.
+## Attribute Reference
```
Comment (Lance52259) [inline docs/data-sources/workspace_flavors]:
```suggestion
* `availability_zone` - (Optional, String) Available partition of the workspace flavors.

## Attribute Reference
```

File: docs/data-sources/workspace_flavors [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+* `region` - (Optional, String) Specifies the region in which to query the data source.
+  If omitted, the provider-level region will be used.
+
+* `vcpus` - (Optional, Int) Specifies  CPU specifications of the workspace flavors.
+
+* `memory` - (Optional, Int) Specifies the memory size of the workspace flavors.
+
+* `os_type` - (Optional, String) The operating system type of the workspace flavors, currently supported: Windows, Linux.  
+  The options are as follows:
+    - **Windows**: The operating system type of the workspace flavors is Windows.
```
Comment (Lance52259) [inline docs/data-sources/workspace_flavors]:
```suggestion
  + **Windows**: The operating system type of the workspace flavors is Windows.
```
the first level of the list using `+` to mark each line and two spaces ahead of them.

File: docs/data-sources/workspace_flavors [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    - **Linux**: The operating system type of the workspace flavors is Linux.
+
+* `availability_zone` - (Optional, String) Available partition of the workspace flavors.
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
```
Comment (Lance52259) [inline docs/data-sources/workspace_flavors]: missing data source ID.

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+			"description":      utils.PathSearch("descriptions", v, nil),
+			"charging_mode":    convertStrValue(utils.PathSearch("charge_mode", v, nil)),
+			"status":           utils.PathSearch("status", v, nil),
+		})
+	}
+	return rst
+}
+
+func convertStrValue(chargeModeCode interface{}) string {
+	if "0" == chargeModeCode {
```
Comment (Lance52259) [inline huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]:
```suggestion
	if chargeModeCode  == "0" {
```

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	curJson := utils.PathSearch("products", resp, make([]interface{}, 0))
+	curArray := curJson.([]interface{})
+	rst := make([]interface{}, 0, len(curArray))
+	for _, v := range curArray {
+		rst = append(rst, map[string]interface{}{
+			"id":               utils.PathSearch("product_id", v, nil),
+			"type":             utils.PathSearch("type", v, nil),
+			"architecture":     utils.PathSearch("architecture", v, nil),
+			"vcpus":            convertStrToInt(utils.PathSearch("cpu", v, nil)),
+			"memory":           utils.UnitConver(utils.PathSearch("memory", v, nil).(string)),
```
Comment (Lance52259) [inline huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: if the result of PathSearch function is nil, that will be panic

File: huaweicloud/utils/utils.go [hunk 6]
```diff
@@ -283,6 +283,17 @@ func GetTimezoneCode() int {
...
+// For example 4096MB = 4GB
+func UnitConver(memory interface{}) int {
+	num, err := strconv.Atoi(memory.(string))
+	if err != nil {
+		fmt.Printf("convert string to int fail")
+	}
+	num /= 1024
+	return num
+}
+
```
Comment (Lance52259) [inline huaweicloud/utils/utils.go]:
```suggestion
// ConvertMemoryUnit is a method that used to convert the memory unit.
// Parameters:
// + memory: Size of the memory in the current unit.
// + diffLevel: Difference level between units before and after conversion.
// diffLevel greater than 0 means that the unit is converted to a higher level, and vice versa to a lower level, e.g.
// the unit of memory input is MB, -2 means it is converted from MB to B, and 2 means it is converted from MB to TB.
func ConvertMemoryUnit(memory interface{}, diffLevel int) int {
	var memoryInt int
	switch memory.(type) {
	case int:
		memoryInt = memory.(int)
	case string:
		var err error
		memoryInt, err = strconv.Atoi(memory.(string))
		if err != nil {
			log.Printf("convert string value (%v) to int fail: %s", memory, err)
		}
	}

	if diffLevel >= 0 {
		return memoryInt / Power(1024, diffLevel)
	}
	return memoryInt * Power(1024, -diffLevel)
}
```
And you should support a function to calculate the power of a number.
```
func Power(base int, exponent int) int {
	if exponent == 0 {
		return 1
	}
	return base * Power(base, exponent-1)
}
```

File: docs/data-sources/workspace_flavors [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+* `system_disk_type` - The workspace flavors system disk type.
+
+* `system_disk_size` - The workspace flavors system disk size in GB.
+
+* `description` - Description of the workspace flavors.
+
+* `charging_mode` - Periodic package identification of the workspace flavors, only on-demand support is available currently.
+  The options are as follows:
+  + **0**: Indicates billing based on package cycle of the workspace flavors.
+  + **1**: Indicates on-demand billing of the workspace.
```
Comment (Lance52259) [inline docs/data-sources/workspace_flavors]:
```suggestion
  + **prePaid**: Indicates billing based on package cycle of the workspace flavors.
  + **postPaid**: Indicates on-demand billing of the workspace.
```

File: docs/data-sources/workspace_flavors (hunk 1 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors]: have changed

File: docs/data-sources/workspace_flavors (hunk 2 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors]: have changed

File: docs/data-sources/workspace_flavors (hunk 3 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors]: have changed

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go (hunk 4 above)
Comment (ruwenqiang123) [reply huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: have changed

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go (hunk 5 above)
Comment (ruwenqiang123) [reply huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: have changed

File: huaweicloud/utils/utils.go (hunk 6 above)
Comment (ruwenqiang123) [reply huaweicloud/utils/utils.go]: have changed

File: docs/data-sources/workspace_flavors (hunk 7 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors]: have changed

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+		}
+		rst = append(rst, v)
+	}
+	return rst
+}
+
+func buildListFlavorsQueryParams(d *schema.ResourceData) string {
+	res := "charge_mode=1"
+	if v, ok := d.GetOk("availability_zone"); ok {
+		res = fmt.Sprintf("%s&availability_zone=%v", res, v)
```
Comment (ShiChangkuo) [inline huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]:
```suggestion
		res += fmt.Sprintf("&availability_zone=%v",  v)
```

File: docs/data-sources/workspace_flavors.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String) Specifies the region in which to query the data source.
+  If omitted, the provider-level region will be used.
+
+* `vcpus` - (Optional, Int) Specifies CPU specification of the workspace flavors.
+
+* `memory` - (Optional, Int) Specifies the memory size of the workspace flavors.
```
Comment (ShiChangkuo) [inline docs/data-sources/workspace_flavors.md]: what's the unit of memory?

File: docs/data-sources/workspace_flavors.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+* `description` - Description of the workspace flavor.
+
+* `charging_mode` - Periodic package identification of the workspace flavor.
+  The valid values are as follows:
+  + **prePaid**: Indicates billing based on package cycle of the workspace flavor.
+  + **postPaid**: Indicates on-demand billing of the workspace flavor.
+
+* `status` - The status of the workspace flavor.  
+  The valid values are as follows:
+  + **normal**: The status is normal of the workspace flavor.
```
Comment (ShiChangkuo) [inline docs/data-sources/workspace_flavors.md:76]: how about we only filter the flavors which are in **normal** status.

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+			"system_disk_size": utils.PathSearch("system_disk_size", v, nil),
+			"description":      utils.PathSearch("descriptions", v, nil),
+			"charging_mode":    convertStrValue(utils.PathSearch("charge_mode", v, nil)),
+			"status":           utils.PathSearch("status", v, nil),
+		})
+	}
+	return rst
+}
+
+func convertStrValue(chargeModeCode interface{}) string {
```
Comment (ShiChangkuo) [inline huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: it's hard to understand from the function name.

```suggestion
func normalizeChargingMode(chargeMode interface{}) string {
```

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go (hunk 8 above)
Comment (ruwenqiang123) [reply huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: have changed

File: docs/data-sources/workspace_flavors.md (hunk 9 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors.md]: have changed

File: docs/data-sources/workspace_flavors.md (hunk 10 above)
Comment (ruwenqiang123) [reply docs/data-sources/workspace_flavors.md:76]: have changed

File: huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go (hunk 11 above)
Comment (ruwenqiang123) [reply huaweicloud/services/workspace/data_source_huaweicloud_workspace_flavors.go]: have changed

File: huaweicloud/services/acceptance/workspace/data_source_huaweicloud_workspace_flavors_test.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  filter_result = [for v in data.huaweicloud_workspace_flavors.filter_by_vcpus.flavors : v.charging_mode == "postPaid"]
+}
+
+output "internal_charging_mode_useful" {
+  value = alltrue(local.filter_result) && length(local.filter_result) > 0
+}
+
+locals {
+  filter_results = [for v in data.huaweicloud_workspace_flavors.filter_by_vcpus.flavors : v.status == "normal"]
+}
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/workspace/data_source_huaweicloud_workspace_flavors_test.go]:
```suggestion
locals {
  vcpus_filter_result = [for v in data.huaweicloud_workspace_flavors.filter_by_vcpus.flavors : v.vcpus == 2]
}

output "is_vcpus_filter_useful" {
  value = alltrue(local.vcpus_filter_result) && length(local.vcpus_filter_result) > 0
}

locals {
  charging_mode_filter_result = [for v in data.huaweicloud_workspace_flavors.filter_by_vcpus.flavors : v.charging_mode == "postPaid"]
}

output "internal_charging_mode_useful" {
  value = alltrue(local.charging_mode_filter_result) && length(local.charging_mode_filter_result) > 0
}

locals {
  status_filter_result = [for v in data.huaweicloud_workspace_flavors.filter_by_vcpus.flavors : v.status == "normal"]
}

Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/workspace/data_source_huaweicloud_workspace_flavors_test.go]: hava changed

**Pull request merged**